source/particles/generators.cc

```diff
@@ -75,6 +142,263 @@ namespace Particles
+          {
+            // Uniform distribution on the interval [0,local_weight_integral).
+            // This will be used to randomly select cells for all local
+            // particles.
+            boost::random::uniform_real_distribution<double>
+              uniform_distribution(0.0, local_weight_integral);
+
+            // Loop over all particles to create locally and pick their cells
+            for (types::particle_index current_particle_index = 0;
+                 current_particle_index < n_local_particles;
+                 ++current_particle_index)
+              {
+                // Draw the random number that determines the cell of the
+                // particle
+                const double random_weight =
+                  uniform_distribution(random_number_generator);
+
+                const std::vector<double>::const_iterator selected_cell =
+                  std::lower_bound(accumulated_cell_weights.begin(),
+                                   accumulated_cell_weights.end(),
+                                   random_weight);
+                const unsigned int cell_index =
+                  std::distance(accumulated_cell_weights.begin(),
+                                selected_cell);
+
+                ++particles_per_cell[cell_index];
+              }
+          }
+        else
+          {
+            // Compute number of particles per cell according to the ratio
+            // between their weight and the local weight integral
+            unsigned int          cell_index        = 0;
+            types::particle_index particles_created = 0;
+
+            for (const auto &cell : triangulation.active_cell_iterators())
+              if (cell->is_locally_owned())
+                {
+                  const types::particle_index particles_to_create =
+                    llround(static_cast<double>(n_local_particles) *
+                            accumulated_cell_weights[cell_index] /
+                            local_weight_integral);
+
+                  // We assume nobody creates more than 4 billion particles in
+                  // one cell
+                  particles_per_cell[cell_index] = static_cast<unsigned int>(
+                    particles_to_create - particles_created);
+                  particles_created += particles_per_cell[cell_index];
+                  ++cell_index;
+                }
+          }
+      }
+
+      // Now generate as many particles per cell as determined above
+      {
+        unsigned int cell_index             = 0;
+        unsigned int current_particle_index = start_particle_id;
+
+        // We first store the generated particles in a vector. Since they are
+        // generated cell-by-cell, they will already be sorted in the correct
+        // order to be later transferred to the multimap with O(N) complexity.
+        // If we would insert them into the multimap one-by-one it would
+        // increase the complexity to O(N log(N)).
```

Comment:
You made me think here. That's because the insertion of the initial few particles doesn't require log(N) steps each, but really just log(n) where n is the number of particles in the map *at the time*, not at the end! So I was wondering whether one might get something better with the naive algorithm -- which I agree we shouldn't use, but I wanted to convince myself that the complexity here is still correct.

And it is: The cost is `sum_n  log(n)`, which is approximately equal to `int_1^N  log(x) dx`, which equates to `1+N log N - N`, which is of course `O(N log N)` as claimed :-)